ast/src/type.rs

```diff
@@ -1,7 +1,76 @@
 use dbg_pls::DebugPls;
+use std::fmt::{Debug, Display, Formatter, Write};
+use std::ops::Deref;
 
 #[derive(Debug, Clone, PartialEq, DebugPls)]
-pub struct Type<'a> {
+pub enum Type<'a> {
+    ///A Simple type with optional parameters (`A`, `A[V]`)
+    Simple(SimpleType<'a>),
+    ///A Lambda declaration with of form `(A, B, ...) => Out`
+    Lambda(LambdaType<'a>),
```

Comment:
Wouldn't `CallableType` be more appropriate here? `Lambda` refers to the actual declaration and not the type.

Methods could also be accepted here, for instance:
```kt
fun count(sized: () => Int) -> Int = $sized()
val digits = List[Int](1, 2, 3)
count($digits.size)
```